cineast-core/src/main/java/org/vitrivr/cineast/core/util/TextRecognizer.java

```diff
@@ -0,0 +1,132 @@
+        );
+        Mat warp = Imgproc.getPerspectiveTransform(src,dst);
+        Mat cropped = new Mat();
+        Imgproc.warpPerspective(frame, cropped, warp, new Size(this.width, this.height));
+        return cropped;
+    }
+
+    /**
+     * rectangularCrop crops the image according to the coordinate supplied
+     * @param coordinate The quadrilateral coordinate of the text within the frame
+     * @param frame The frame containing the text
+     * @return The rectangularly cropped image containing the text
+     */
+    private Mat rectangularCrop(Point[] coordinate, Mat frame) {
+        Rect rectangle = Imgproc.boundingRect(Converters.vector_Point_to_Mat(Arrays.asList(coordinate)));
+        if (rectangle.x+rectangle.width > frame.width()) {
+            rectangle.width = frame.width() - rectangle.x;
+        }
+        if (rectangle.y+rectangle.height > frame.height()) {
+            rectangle.height = frame.height() - rectangle.y;
+        }
+        Mat cropped = new Mat(frame, rectangle);
+        Mat resized = new Mat();
+        Imgproc.resize(cropped, resized, new Size(this.width, this.height));
+        return resized;
+    }
+
+    /**
+     * decode takes an output layer and extracts the string using CTC-based decoding
+     * @param predictions The output layer of the model
+     * @return The extracted string from the layer
+     */
+    private String decode(Mat predictions) {
+        List<String> text = new ArrayList<>();
+
+        for (int i=0; i<predictions.rows(); i++) {
+            int maxIndex = 0;
+            for (int j=1; j<predictions.cols(); j++) {
+                if (predictions.get(i,j)[0] > predictions.get(i,maxIndex)[0]) {
+                    maxIndex = j;
+                }
+            }
+            if (maxIndex > 0 && maxIndex <= this.alphabet.length) {
+                text.add(this.alphabet[maxIndex-1]);
+            } else {
+                text.add("-");
+            }
+        }
+        StringBuffer char_list = new StringBuffer();
+        for (int i = 0; i<text.size(); i++) {
+            if (!text.get(i).equals("-") && !(i > 0 && text.get(i).equals(text.get(i-1)))) {
+                char_list.append(text.get(i));
+            }
+        }
+        return char_list.toString();
+    }
+
+    /**
+     *
+     * @param coordinate The coordinate of the text
+     * @param frame The frame cotnaining the text
+     * @param quadrilateral If true, the region of the coordinates is regarded as quadrilateral, and quadrilateral cropping is applied
+     * @return The extracted text within the coordinate.
+     */
+    public String recognize(Point[] coordinate, Mat frame, boolean quadrilateral) {
+        Mat cropped = quadrilateral ? quadrilateralCrop(coordinate, frame) : rectangularCrop(coordinate, frame);
+        Imgproc.cvtColor(cropped, cropped, Imgproc.COLOR_RGB2GRAY);
+        Mat blob = Dnn.blobFromImage(cropped, 1/127.5, new Size(100, 32), new Scalar(127.5));
```

Comment:
What is the meaning behind the 127.5?

Reply:
This is the scaling factor, which every RGB value will be multiplied by in order to scale the input image to a particular range that is needed for this architecture.